libbpf-tools/gethostlatency.bpf.c

```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2021 Hengqi Chen
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "gethostlatency.h"
+
+#define MAX_ENTRIES 10240
+
+const volatile pid_t targ_tgid = 0;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32);
+	__type(value, struct val_t);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+static __always_inline
+int probe_entry(struct pt_regs *ctx) {
+	if (!PT_REGS_PARM1(ctx))
+		return 0;
+
+	struct val_t val = {};
+	u32 pid = bpf_get_current_pid_tgid();
+
+	if (targ_tgid && targ_tgid != pid)
+		return 0;
+
+	if (bpf_get_current_comm(&val.comm, sizeof(val.comm)) == 0) {
+		bpf_probe_read_user(&val.host, sizeof(val.host),
+					   (void *)PT_REGS_PARM1(ctx));
+		val.pid = bpf_get_current_pid_tgid();
+		val.ts = bpf_ktime_get_ns();
+		bpf_map_update_elem(&start, &pid, &val, BPF_ANY);
+	}
+
+	return 0;
+}
+
+static __always_inline
+int probe_return(struct pt_regs *ctx) {
+	struct val_t *valp;
+	struct data_t data = {};
+	u64 delta;
+	u32 pid = bpf_get_current_pid_tgid();
+	u64 tsp = bpf_ktime_get_ns();
+
+	valp = bpf_map_lookup_elem(&start, &pid);
+	if (valp == 0)
+		return 0;
+
+	bpf_probe_read_kernel(&data.comm, sizeof(data.comm), valp->comm);
+	bpf_probe_read_kernel(&data.host, sizeof(data.host), (void *)valp->host);
+	data.pid = valp->pid;
+	data.delta = tsp - valp->ts;
+	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU, &data,
+			sizeof(data));
+	bpf_map_delete_elem(&start, &pid);
+	return 0;
+}
+
+SEC("uprobe/c/getaddrinfo")
+int BPF_PROG(getaddrinfo_entry)
+{
+	probe_entry((void *)ctx);
+	return 0;
+}
+
+SEC("uprobe/c/gethostbyname")
+int BPF_PROG(gethostbyname_entry)
+{
+	probe_entry((void *)ctx);
+	return 0;
+}
+
+SEC("uprobe/c/gethostbyname2")
+int BPF_PROG(gethostbyname2_entry)
+{
+	probe_entry((void *)ctx);
+	return 0;
+}
+
+SEC("uretprobe/c/getaddrinfo")
+int BPF_PROG(getaddrinfo_return)
+{
+	probe_return((void *)ctx);
+	return 0;
+}
+
+SEC("uretprobe/c/gethostbyname")
+int BPF_PROG(gethostbyname_return)
+{
+	probe_return((void *)ctx);
+	return 0;
+}
+
+SEC("uretprobe/c/gethostbyname2")
+int BPF_PROG(gethostbyname2_return)
```

Comment:
you don't really need all these program copies, it's enough to have one uprobe and one uretprobe program and attach them to multiple functions at runtime.

also BPF_PROG is not intended for uprobe/uretprobe, it's for fentry/fexit and alike, so please just do

```
int hanlde_entry(struct pt_regs *ctx) { ... }
```

Reply:
Thanks for the insight.

Reply:
> it's enough to have one uprobe and one uretprobe program and attach them to multiple functions at runtime

@anakryiko @chenhengqi I ran `gethostlatency` under ASan and it reported several memory leaks:
```
==14089==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 56 byte(s) in 1 object(s) allocated from:
    #0 0x7f03aa453527 in __interceptor_calloc (/lib64/libasan.so.6+0xab527)
    #1 0x46c65d in bpf_program__attach_perf_event_opts /home/vagrant/bcc/src/cc/libbpf/src/libbpf.c:9163
    #2 0x470cca in bpf_program__attach_uprobe_opts /home/vagrant/bcc/src/cc/libbpf/src/libbpf.c:9668
    #3 0x471271 in bpf_program__attach_uprobe /home/vagrant/bcc/src/cc/libbpf/src/libbpf.c:9699
    #4 0x405f97 in attach_uprobes /home/vagrant/bcc/libbpf-tools/gethostlatency.c:162
    #5 0x406db3 in main /home/vagrant/bcc/libbpf-tools/gethostlatency.c:253
    #6 0x7f03a98651e1 in __libc_start_main (/lib64/libc.so.6+0x281e1)

Direct leak of 56 byte(s) in 1 object(s) allocated from:
    #0 0x7f03aa453527 in __interceptor_calloc (/lib64/libasan.so.6+0xab527)
    #1 0x46c65d in bpf_program__attach_perf_event_opts /home/vagrant/bcc/src/cc/libbpf/src/libbpf.c:9163
    #2 0x470cca in bpf_program__attach_uprobe_opts /home/vagrant/bcc/src/cc/libbpf/src/libbpf.c:9668
    #3 0x471271 in bpf_program__attach_uprobe /home/vagrant/bcc/src/cc/libbpf/src/libbpf.c:9699
    #4 0x40622c in attach_uprobes /home/vagrant/bcc/libbpf-tools/gethostlatency.c:176
    #5 0x406db3 in main /home/vagrant/bcc/libbpf-tools/gethostlatency.c:253
    #6 0x7f03a98651e1 in __libc_start_main (/lib64/libc.so.6+0x281e1)

Direct leak of 56 byte(s) in 1 object(s) allocated from:
    #0 0x7f03aa453527 in __interceptor_calloc (/lib64/libasan.so.6+0xab527)
    #1 0x46c65d in bpf_program__attach_perf_event_opts /home/vagrant/bcc/src/cc/libbpf/src/libbpf.c:9163
    #2 0x470cca in bpf_program__attach_uprobe_opts /home/vagrant/bcc/src/cc/libbpf/src/libbpf.c:9668
    #3 0x471271 in bpf_program__attach_uprobe /home/vagrant/bcc/src/cc/libbpf/src/libbpf.c:9699
    #4 0x40643e in attach_uprobes /home/vagrant/bcc/libbpf-tools/gethostlatency.c:184
    #5 0x406db3 in main /home/vagrant/bcc/libbpf-tools/gethostlatency.c:253
    #6 0x7f03a98651e1 in __libc_start_main (/lib64/libc.so.6+0x281e1)

Direct leak of 56 byte(s) in 1 object(s) allocated from:
    #0 0x7f03aa453527 in __interceptor_calloc (/lib64/libasan.so.6+0xab527)
    #1 0x46c65d in bpf_program__attach_perf_event_opts /home/vagrant/bcc/src/cc/libbpf/src/libbpf.c:9163
    #2 0x470cca in bpf_program__attach_uprobe_opts /home/vagrant/bcc/src/cc/libbpf/src/libbpf.c:9668
    #3 0x471271 in bpf_program__attach_uprobe /home/vagrant/bcc/src/cc/libbpf/src/libbpf.c:9699
    #4 0x405d85 in attach_uprobes /home/vagrant/bcc/libbpf-tools/gethostlatency.c:154
    #5 0x406db3 in main /home/vagrant/bcc/libbpf-tools/gethostlatency.c:253
    #6 0x7f03a98651e1 in __libc_start_main (/lib64/libc.so.6+0x281e1)

SUMMARY: AddressSanitizer: 224 byte(s) leaked in 4 allocation(s).
```
Looks like either `libbpf` can't free the memory allocated by `bpf_program__attach_uprobe` or `bpf_program__attach_uprobe` isn't used correctly.

Reply:
@evverx Thanks for the report. Will take a look at this.